engine/src/main/java/org/hibernate/validator/internal/engine/constraintvalidation/ConstraintTree.java

```diff
@@ -120,7 +121,13 @@ private ValidationException getExceptionForNullValidator(Type validatedValueType
 					validator = constraintValidatorForDefaultConstraintValidatorFactory;
 					if ( validator == null ) {
 						validator = getInitializedConstraintValidator( validationContext );
-						constraintValidatorForDefaultConstraintValidatorFactory = validator;
+
+						// HibernateConstraintValidator instances should not be cached within this constraint tree instance, only ConstraintValidatorManager
+						// should handle their caching. That's because the HibernateConstraintValidatorInitializationContext influences their caching behaviour.
+						// (HibernateConstraintValidatorInitializationContext could have changed during a ValidatorFactory lifetime - via usingContext())
+						if ( !( validator instanceof HibernateConstraintValidator ) ) {
+							constraintValidatorForDefaultConstraintValidatorFactory = validator;
+						}
```

Comment:
This I don't like. The fact is that in the end I would like to have the `HibernateConstraintValidator` `initialize()` contract moved to `ConstraintValidator` and that would defeat the purpose of this optimization.

Thus I think we should include the initialization context right now in all the caching mechanisms.

Will give it a try.